ipld/blockstore/src/lib.rs

```diff
@@ -34,17 +76,7 @@ pub trait Blockstore {
     }
 
     /// Puts the block into the blockstore, computing the hash with the specified multicodec.
-    ///
-    /// By default, this defers to put.
-    fn put<D>(&self, mh_code: multihash::Code, block: &Block<D>) -> Result<Cid>
-    where
-        Self: Sized,
-        D: AsRef<[u8]>,
-    {
-        let k = block.cid(mh_code);
-        self.put_keyed(&k, block.as_ref())?;
-        Ok(k)
-    }
+    fn put(&self, mh_code: u64, block: &dyn Block) -> Result<Cid>;
```

Comment:
- I'm using a trait for `Block` to allow abstracting over blocks that own their data, and blocks that borrow their data.
- After implementing this, I think we need to replace `mh_code` with some form of `MultihashType { size: u32, code: u64 }`.

Reply:
Alternatively, we _could_ simply remove the multihash-code parameter and say that the blockstore gets to decide.

Reply:
I don't think removing it is right at this base level. Currying it in to a wrapper store could make sense though.